pgjdbc/src/test/java/org/postgresql/test/jdbc2/TimezoneCachingTest.java

```diff
@@ -0,0 +1,410 @@
+      rs.getInt(1);
+      assertEquals("Cache never initialized: must be null", null, getTimeZoneCache(rs));
+      rs.getDate(2);
+      assertEquals("Cache initialized by getDate(x): must not be null", tz, getTimeZoneCache(rs));
+      rs.close();
+    } finally {
+      TestUtil.closeQuietly(rs);
+      TestUtil.closeQuietly(pstmt);
+      TestUtil.closeQuietly(stmt);
+    }
+  }
+
+  /**
+   * Test to check the internal cached timezone of a result set is used as expected.
+   */
+  public void testResultSetCachedTimezoneUsage() throws SQLException {
+    Timestamp ts = new Timestamp(2016 - 1900, 1 - 1, 31, 0, 0, 0, 0);
+    Statement stmt = null;
+    PreparedStatement pstmt = null;
+    ResultSet rs = null;
+    TimeZone tz1 = TimeZone.getTimeZone("GMT+8:00");
+    TimeZone tz2 = TimeZone.getTimeZone("GMT-2:00");
+    Calendar c1 = new GregorianCalendar(tz1);
+    Calendar c2 = new GregorianCalendar(tz2);
+    try {
+      TimeZone.setDefault(tz1);
+      pstmt = con.prepareStatement("INSERT INTO testtz VALUES (?,?)");
+      pstmt.setInt(1, 1);
+      // We are in tz1, so timestamp added as tz1.
+      pstmt.setTimestamp(2, ts);
+      pstmt.addBatch();
+      pstmt.executeBatch();
+      stmt = con.createStatement();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertTrue(
+          "Current TZ is tz1, empty cache to be initialized to tz1 => retrieve in tz1, timestamps must be equal",
+          rs.getTimestamp(2).equals(ts));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      TimeZone.setDefault(tz2);
+      assertTrue(
+          "Current TZ is tz2, empty cache to be initialized to tz2 => retrieve in tz2, timestamps cannot be equal",
+          !rs.getTimestamp(2).equals(ts));
+      assertTrue(
+          "Explicit tz1 calendar, so timestamps must be equal",
+          rs.getTimestamp(2, c1).equals(ts));
+      assertTrue(
+          "Cache was initialized to tz2, so timestamps cannot be equal",
+          !rs.getTimestamp(2).equals(ts));
+      TimeZone.setDefault(tz1);
+      assertTrue(
+          "Cache was initialized to tz2, so timestamps cannot be equal",
+          !rs.getTimestamp(2).equals(ts));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertTrue(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          !rs.getTimestamp(2, c2).equals(ts));
+      assertTrue(
+          "Current TZ is tz1, empty cache to be initialized to tz1 => retrieve in tz1, timestamps must be equal",
+          rs.getTimestamp(2).equals(ts));
+      assertTrue(
+          "Explicit tz2 calendar, so timestamps cannot be equal",
+          !rs.getTimestamp(2, c2).equals(ts));
+      assertTrue(
+          "Explicit tz2 calendar, so timestamps must be equal",
+          rs.getTimestamp(2, c1).equals(ts));
```

Comment:
Use `assertEquals`, and know which outcome is expected.
At the end of the day, it is a test case in a controlled environment, so it should always return some known-ahead result.